test/e2e/kube-ovn/underlay/underlay.go

```diff
@@ -417,4 +435,307 @@ var _ = framework.Describe("[group:underlay]", func() {
+
+		ginkgo.By("step4: check if kube-ovn-controller restart")
+		restartCmd := "kubectl rollout restart deployment kube-ovn-controller -n kube-system"
+		_, err = exec.Command("bash", "-c", restartCmd).CombinedOutput()
+		framework.ExpectNoError(err, "restart kube-ovn-controller")
+		checkU2OItems(true, subnet, underlayPod, overlayPod)
+
+		ginkgo.By("step5: Disable u2o check after restart kube-controller")
+		podClient.DeleteSync(u2oPodNameUnderlay)
+
+		subnet = subnetClient.Get(subnetName)
+		modifiedSubnet = subnet.DeepCopy()
+		modifiedSubnet.Spec.U2OInterconnection = false
+		subnetClient.PatchSync(subnet, modifiedSubnet)
+		time.Sleep(5 * time.Second)
+		underlayPod = podClient.CreateSync(originUnderlayPod)
+		subnet = subnetClient.Get(subnetName)
+		checkU2OItems(false, subnet, underlayPod, overlayPod)
+
+		ginkgo.By("step6: recover enable u2o check after restart kube-controller")
+		podClient.DeleteSync(u2oPodNameUnderlay)
+
+		subnet = subnetClient.Get(subnetName)
+		modifiedSubnet = subnet.DeepCopy()
+		modifiedSubnet.Spec.U2OInterconnection = true
+		subnetClient.PatchSync(subnet, modifiedSubnet)
+		time.Sleep(5 * time.Second)
+		underlayPod = podClient.CreateSync(originUnderlayPod)
+		subnet = subnetClient.Get(subnetName)
+		checkU2OItems(true, subnet, underlayPod, overlayPod)
+	})
 })
+
+func checkU2OItems(isEnableU2O bool, subnet *apiv1.Subnet, underlayPod, overlayPod *corev1.Pod) {
+
+	ginkgo.By("checking underlay subnet's u2o interconnect ip")
+	if isEnableU2O {
+		framework.ExpectEqual(subnet.Spec.U2OInterconnection, true)
+		framework.ExpectNotEqual(subnet.Status.U2OInterconnectionIP, "")
+	} else {
+		framework.ExpectNotEqual(subnet.Spec.U2OInterconnection, true)
+		framework.ExpectEqual(subnet.Status.U2OInterconnectionIP, "")
+	}
+
+	v4gw, v6gw := util.SplitStringIP(subnet.Spec.Gateway)
+
+	underlayCidr := strings.Split(subnet.Spec.CIDRBlock, ",")
+	for _, cidr := range underlayCidr {
+		if util.CheckProtocol(cidr) == apiv1.ProtocolIPv4 {
+			ginkgo.By("checking underlay subnet's using ips")
+			if isEnableU2O {
+				framework.ExpectEqual(int(subnet.Status.V4UsingIPs), 2)
+			} else {
+				framework.ExpectEqual(int(subnet.Status.V4UsingIPs), 1)
+			}
+
+			agName := strings.Replace(fmt.Sprintf("%s.u2o_exclude_ip.ip4", subnet.Name), "-", ".", -1)
+			ginkgo.By("checking underlay subnet's policy1 route")
+			hitPolicyStr := fmt.Sprintf("%d ip4.dst == $%s && ip4.src == %s allow", util.SubnetRouterPolicyPriority, agName, cidr)
+			checkPolicy(hitPolicyStr, isEnableU2O)
+
+			ginkgo.By("checking underlay subnet's policy2 route")
+			hitPolicyStr = fmt.Sprintf("%d ip4.dst == %s && ip4.dst != $%s allow", util.SubnetRouterPolicyPriority, cidr, agName)
+			checkPolicy(hitPolicyStr, isEnableU2O)
+
+			ginkgo.By("checking underlay subnet's policy3 route")
+			hitPolicyStr = fmt.Sprintf("%d ip4.src == %s reroute %s", util.GatewayRouterPolicyPriority, cidr, v4gw)
+			checkPolicy(hitPolicyStr, isEnableU2O)
+		} else {
```

Comment:
Almost the same code as the IPv6 check. It would be better to simplify these codes.